### PR TITLE
Initial coveo.analytics.js project

### PR DESCRIPTION
**Pull request by pastjean**
Initial project with the basic analytics module (the module to push events into the api).
There is also a simple API to call the same style googleanalytics does with its analytics.js library

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    "es6-promise": "^3.1.2",
+    "exports-loader": "^0.6.3",
+    "isomorphic-fetch": "^2.2.1",
+    "ts-loader": "^0.8.2",
+    "tsc": "^1.20150623.0",
+    "tslint": "^3.7.3",
+    "tslint-loader": "^2.1.3",
+    "typescript": "^1.8.9",
+    "typings": "^0.7.12",
+    "webpack": "^1.12.15"
```
Comment (GermainBergeron) [inline package.json]: In the JsAdmin we had a lot of problem when we didn't use fixed version. Consider removing the `^` for every dependencies

File: .editorconfig [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+charset = utf-8
+
+[*.js]
+charset = utf-8
+indent_style = space
+indent_size = 4
+
+[{package.json,.travis.yml}]
+indent_style = space
+indent_size = 2
```
Comment (wfortin) [inline .editorconfig]: charset = utf-8 should be on [*]

File: .travis.yml [hunk 3]
```diff
@@ -0,0 +1,23 @@
+sudo: false
+language: node_js
+node_js:
+- stable
+install:
+- npm install
+- "$(npm bin)/typings install"
```
Comment (wfortin) [inline .travis.yml]: might want to make a `postinstall` script. This will only require to run `npm install`.

File: .travis.yml [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+script:
+- npm run-script build:tsc
+- npm run-script build:webpack
+deploy:
+  skip_cleanup: true
+  provider: npm
+  email: pa@stjean.me
```
Comment (wfortin) [inline .travis.yml]: coveo email ?

File: README.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+[![Build Status](https://travis-ci.org/coveo/coveo.analytics.js.svg?branch=master)](https://travis-ci.org/coveo/coveo.analytics.js)
+[![dependency status](https://david-dm.org/coveo/coveo.analytics.js.svg)](https://david-dm.org/coveo/coveo.analytics.js)
+[![dev dependency status](https://david-dm.org/coveo/coveo.analytics.js/dev-status.svg)](https://david-dm.org/coveo/coveo.analytics.js#info=devDependencies)
+
+> Coveo's usage analytics javascript client
+
+## Usage
+
+You have to provide your own `promise` and `fetch` api compatible libraries.
```
Comment (wfortin) [inline README.md]: API in caps.

File: README.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+npm run build:webpack
+# code code code
+# open pull request
+```
+
+## License
+
+MIT license; see [LICENSE](./LICENSE).
+
+[![forthebadge](http://forthebadge.com/images/badges/gluten-free.svg)](http://forthebadge.com)
```
Comment (wfortin) [inline README.md]: WUT ?

File: src/SimpleAnalytics.ts [hunk 7]
```diff
@@ -0,0 +1,63 @@
+import * as analytics from './analytics';
+import objectAssign from './objectAssign';
+
+// SimpleAPI permits to mimic the GoogleAnalytics simple to call api, but we
+// send events to Coveo usage analytics service usign thoses calls.
```
Comment (GermainBergeron) [inline src/SimpleAnalytics.ts]: // SimpleAPI mimics GoogleAnalytics simple to call api, but we
// send events to Coveo usage analytics service using those calls.

File: src/SimpleAnalytics.ts [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    // @param token is your coveo access_token / api_key / ...
+    // @param endpoint is the endpoint you want to target defaults to the
+    //        usage analytics production endpoint
+    init(token: string, endpoint: string = analytics.Endpoints.default ) {
+        if (token === undefined) {
+            throw new Error(`You must pass your token when you call 'init'`);
+        }
+
+        this.isInited = true;
+        this.client = new analytics.Client({
```
Comment (GermainBergeron) [inline src/SimpleAnalytics.ts]: You could check if `this.client` is defined instead of having a boolean `isInited`

File: src/SimpleAnalytics.ts [hunk 9]
```diff
@@ -0,0 +1,63 @@
+import * as analytics from './analytics';
+import objectAssign from './objectAssign';
+
+// SimpleAPI permits to mimic the GoogleAnalytics simple to call api, but we
+// send events to Coveo usage analytics service usign thoses calls.
+class SimpleAPI {
+    private client: analytics.Client;
+    private isInited: boolean = false;
```
Comment (wfortin) [inline src/SimpleAnalytics.ts]: isInitiated ?

File: src/SimpleAnalytics.ts [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+    // init initializes a new SimpleAPI client.
+    // @param token is your coveo access_token / api_key / ...
+    // @param endpoint is the endpoint you want to target defaults to the
+    //        usage analytics production endpoint
+    init(token: string, endpoint: string = analytics.Endpoints.default ) {
+        if (token === undefined) {
```
Comment (wfortin) [inline src/SimpleAnalytics.ts]: `if (typeof myVar != 'undefined')` is the way to go actualy.

File: src/SimpleAnalytics.ts [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+        this.isInited = true;
+        this.client = new analytics.Client({
+            token: token,
+            endpoint: endpoint
+        });
+    }
+
+    send(event: string, customData: any) {
+        if (!this.isInited) {
+            console.error(`You must init before sending an event`);
```
Comment (wfortin) [inline src/SimpleAnalytics.ts]: in `init` you throw, but in `send` you log. We should use the same logic on both i think.

File: src/SimpleAnalytics.ts [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+                language: navigator.language,
+                title: document.title,
+                customData: customData
+            });
+            return;
+        default:
+            console.error(`Event type: '${event}' not implemented`);
+            return;
+        }
+    }
```
Comment (wfortin) [inline src/SimpleAnalytics.ts]: The `switch` formatting is weird here.

File: src/analytics.ts [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+            }).then((response) => {
+            return response.json().then((data) => {
+                return {
+                    raw: response,
+                    visitId: data['visitId'],
+                    visitorId: data['visitorId']
+                };
+            });
+        });
+    }
```
Comment (GermainBergeron) [inline src/analytics.ts]: Those last 3 functions look a lot like each others. You should probably have some default options and response function

File: src/SimpleAnalytics.ts [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+            return;
+        }
+
+        customData = objectAssign({}, {
+            hash: window.location.hash
+        }, customData);
+
+        switch (event) {
+        case 'pageview':
+            this.client.SendViewEvent({
```
Comment (wfortin) [inline src/SimpleAnalytics.ts]: I'm just being picky here but methods should start with a lowercase
```this.client.sendViewEvent({})```

File: src/analytics.ts [hunk 15]
```diff
@@ -0,0 +1,192 @@
+
+export const Endpoints = {
+  default: 'https://usageanalytics.coveo.com/rest/v15',
+  production: 'https://usageanalytics.coveo.com/rest/v15',
+  dev: 'https://usageanalyticsdev.coveo.com/rest/v15',
+  staging: 'https://usageanalyticsstaging.coveo.com/rest/v15'
```
Comment (wfortin) [inline src/analytics.ts]: Should we have a parameter for the version ?

File: src/analytics.ts [hunk 16]
```diff
@@ -0,0 +1,192 @@
...
+    GetHealth(): Promise<HealthResponse> {
+        return fetch(this.endpoint + '/analytics/monitoring/health').then((response) => {
+            return response.json().then((data) => {
+                return {
+                    raw: response,
+                    status: data['status']
+                };
+            });
+        });
+    }
```
Comment (wfortin) [inline src/analytics.ts]: As I said earlier: each methods should start with a lowercase.

File: src/analytics.ts [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+    private endpoint: string;
+    private token: string;
+
+    constructor(opts: ClientOptions = {token: ''} ) {
+        this.endpoint = opts.endpoint || Endpoints.default;
+        this.token = opts.token;
+    }
+
+    SendSearchEvent(request: SearchEventRequest): Promise<SearchEventResponse> {
+        return fetch(this.endpoint + '/analytics/search', {
```
Comment (wfortin) [inline src/analytics.ts]: YEAH ⚡️

File: src/analytics.ts [hunk 18]
```diff
@@ -0,0 +1,192 @@
...
+        });
+    }
+    SendClickEvent(request: ClickEventRequest): Promise<ClickEventResponse> {
+        return fetch(this.endpoint + '/analytics/click', {
+                method: 'POST',
+                headers: {authorization: `BEARER ${this.token}`},
+                mode: 'cors',
+                body: JSON.stringify(request)
+            }).then((response) => {
+            return response.json().then((data) => {
```
Comment (wfortin) [inline src/analytics.ts]: interface for data ?

you will then be able to do data.visitId. `data['visitId']` annoys me 😛

File: src/analytics.ts [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+    location: string;
+    referrer?: string;
+    title?: string;
+}
+
+export interface DefaultEventResponse {
+    raw: IResponse;
+    visitId: string;
+    visitorId: string;
+}
```
Comment (wfortin) [inline src/analytics.ts]: This is pure awesomeness!

File: src/analytics.ts [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+
+export interface SearchEventResponse extends DefaultEventResponse {}
+export interface ClickEventResponse extends DefaultEventResponse {}
+export interface CustomEventResponse extends DefaultEventResponse {}
+export interface ViewEventResponse extends DefaultEventResponse {}
+
+export interface VisitResponse {
+    raw: IResponse;
```
Comment (wfortin) [inline src/analytics.ts]: No `I` prefix in interfaces, https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines#names

File: src/browser.ts [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+import * as _analytics from './index';
+
+declare const global: any;
+
+// CoveoUAGlobal is the interface for the global function which also has a
+// queue `q` of unexecuted parameters
+export interface CoveoUAGlobal {
+  (action: string, ...params: Array<string>): void;
+  // CoveoAnalytics.q ; // is the queue of last called actions before lib was included
+  q?: Array<Array<string>>;
```
Comment (wfortin) [inline src/browser.ts]: `queue` ? we don't win much for truncating variable names.

File: src/browser.ts [hunk 22]
```diff
@@ -0,0 +1,32 @@
+import SimpleAnalytics from './SimpleAnalytics';
+import * as _analytics from './index';
```
Comment (wfortin) [inline src/browser.ts]: Why the _analytics ?

File: src/index.ts [hunk 23]
```diff
@@ -0,0 +1,5 @@
+import * as _analytics from './analytics';
+import * as _SimpleAnalytics from './SimpleAnalytics';
+
+export const analytics = _analytics;
+export const SimpleAnalytics = _SimpleAnalytics;
```
Comment (wfortin) [inline src/index.ts]: You should be able to do something like this:
```typescript
export { * as t1 } from './test1';
export { * as t2 } from './test2';
```

File: src/objectAssign.ts [hunk 24]
```diff
@@ -0,0 +1,38 @@
...
+                    output[symbol] = from[symbol];
+                }
+            });
+        }
+    });
+    return output;
+};
+
+export const assign: IObjectAssign = typeof (<any>Object).assign === 'function' ?  (<any>Object).assign : objectAssignPonyfill;
+export default assign;
```
Comment (wfortin) [inline src/objectAssign.ts]: My want to state the source if you found that on the web.

File: tsconfig.json [hunk 25]
```diff
@@ -0,0 +1,18 @@
+{
+  "compilerOptions": {
+    "target": "es5",
+    "outDir": "dist",
+    "module": "commonjs",
+    "declaration": true,
+    "noImplicitAny": true,
```
Comment (wfortin) [inline tsconfig.json]: NICE !

File: tslint.json [hunk 26]
```diff
@@ -0,0 +1,63 @@
...
+    "whitespace": [
+      true,
+      "check-branch",
+      "check-decl",
+      "check-operator",
+      "check-separator",
+      "check-type"
+    ]
+  }
+}
```
Comment (wfortin) [inline tslint.json]: NICE !

File: webpack.config.js [hunk 27]
```diff
@@ -0,0 +1,35 @@
...
+    },
+    plugins:[
+        new webpack.ProvidePlugin({
+            'Promise': 'exports?global.Promise!es6-promise',
+            'window.fetch': 'exports?self.fetch!isomorphic-fetch',
+        })
+    ],
+    ts: {
+        compilerOptions: {
+            "declaration": false,
```
Comment (wfortin) [inline webpack.config.js]: We should enable the declarations if someone wants to use the lib in TS? or he will include directly the ts files ?

File: package.json (hunk 1 above)
Comment (wfortin) [inline package.json]: We should take a look at https://docs.npmjs.com/cli/shrinkwrap one day 😄

File: webpack.config.js (hunk 27 above)
Comment (pastjean) [inline webpack.config.js]: Webpack is only used to make a web bundle! 
I use basic tsc to generate the declaration, try npm install coveo.analytics for a sneak peak

File: src/index.ts (hunk 23 above)
Comment (pastjean) [inline src/index.ts]: Don't remember trying this one but I tried a lot of combinations. I'll try again

File: src/browser.ts (hunk 21 above)
Comment (pastjean) [inline src/browser.ts]: that was to do the same as google analytics

File: src/browser.ts (hunk 22 above)
Comment (pastjean) [inline src/browser.ts]: relics from the past 👀 
where i was exporting with the same name

File: src/analytics.ts (hunk 16 above)
Comment (pastjean) [inline src/analytics.ts]: will do

File: src/analytics.ts (hunk 18 above)
Comment (pastjean) [inline src/analytics.ts]: Promise<SearchEventResponse>, if it doesnt match the interface there's an error. but i could probably make it explicit

File: src/analytics.ts (hunk 15 above)
Comment (pastjean) [inline src/analytics.ts]: um... :\ quite practical to test agains another version without changing 2-3 parameters

File: src/SimpleAnalytics.ts (hunk 12 above)
Comment (pastjean) [inline src/SimpleAnalytics.ts]: ill ping you for that one

File: src/SimpleAnalytics.ts (hunk 11 above)
Comment (pastjean) [inline src/SimpleAnalytics.ts]: yup true. I'll do this

File: .travis.yml (hunk 4 above)
Comment (pastjean) [inline .travis.yml]: actually i think this is optionnal, my npm account is binded to this email tough. ill fix this

File: src/analytics.ts (hunk 20 above)
Comment (pastjean) [inline src/analytics.ts]: sarry that comes from the promise.d.ts

File: src/index.ts (hunk 23 above)
Comment (pastjean) [inline src/index.ts]: ouin j'avais testé ça et ça ne fonctionne pas tu peux pas namespacer un export en important en meme temps, j'ai opté pour ça a la place 
```typescript
import * as analytics from './analytics';
import * as SimpleAnalytics from './SimpleAnalytics';

export {
    analytics,
    SimpleAnalytics
}
```

File: src/SimpleAnalytics.ts [hunk 28]
```diff
@@ -0,0 +1,58 @@
...
+                location: window.location.toString(),
+                referrer: document.referrer,
+                language: navigator.language,
+                title: document.title,
+                customData: customData
+            });
+            return;
+        default:
+            throw new Error(`Event type: '${event}' not implemented`);
+        }
```
Comment (wfortin) [inline src/SimpleAnalytics.ts]: Formatting is still weird here.

File: src/analytics.ts (hunk 15 above)
Comment (wfortin) [inline src/analytics.ts]: I mean like 
```typescript

export const version = 'v15';

export const Endpoints = {
   default: 'https://usageanalytics.coveo.com/rest/',
   production: 'https://usageanalytics.coveo.com/rest/',
   dev: 'https://usageanalyticsdev.coveo.com/rest/',
 +  staging: 'https://usageanalyticsstaging.coveo.com/rest/'
```

Comment (wfortin) [inline src/analytics.ts]: Not sure though...

File: src/analytics.ts [hunk 29]
```diff
@@ -0,0 +1,142 @@
...
+        return this.sendEvent('view', request).then(defaultResponseTransformer);
+    }
+    getVisit(): Promise<VisitResponse> {
+        return fetch(this.endpoint + '/analytics/visit')
+            .then(defaultResponseTransformer);
+    }
+    getHealth(): Promise<HealthResponse> {
+        return fetch(this.endpoint + '/analytics/monitoring/health')
+            .then(defaultResponseTransformer);
+    }
```
Comment (wfortin) [inline src/analytics.ts]: GRRRRREAT!

File: src/analytics.ts (hunk 20 above)
Comment (wfortin) [inline src/analytics.ts]: 😢

File: src/browser.ts (hunk 21 above)
Comment (wfortin) [inline src/browser.ts]: Keep it like that then! 😄

File: src/objectAssign.ts [hunk 30]
```diff
@@ -0,0 +1,40 @@
+// Comes from the javascrpt version of objectAssing ponyfill by sindresorhus
```
Comment (wfortin) [inline src/objectAssign.ts]:
```
// objectAssing ponyfill by sindresorhus ported to typescript
// source: https://github.com/sindresorhus/object-assign/blob/master/index.js
```

**Comment by wfortin** [comment]
Added some small comments after your last update, awesome job man! Fix those then 🚢  it!

File: .travis.yml (hunk 3 above)
Comment (pastjean) [inline .travis.yml]: le post install script roule toujours. meme pour les users qui vont installer mon package en dépendances (alors qu'ils en on pas besoin) typings est une dev dependency.

File: README.md [hunk 31]
```diff
@@ -0,0 +1,80 @@
...
+<script>
+(function(k,r,y,p,t,o,n){
+n='coveoua';k[n]=k[n]||function(){(k[n].q=k[n].q|| []).push(arguments)};
+k[n].t=Date.now();o=r.createElement(y);o.async=1;o.src=p;
+t=r.getElementsByTagName(y)[0];t.parentNode.insertBefore(o,t)
+})(window,document,'script','https://static.cloud.coveo.com/coveo.analytics.js/coveo.analytics.js')
+
+// Replace YOUR-TOKEN with your real token
+// (eg: an API key which has the rights to write into Coveo UsageAnalytics)
+coveoua('init','YOUR-TOKEN');
```
Comment (GuiSim) [inline README.md]: Je mettrais `'YOUR-TOKEN'` partout où on doit remplacer le token. Consistency is good.

File: src/SimpleAnalytics.ts [hunk 32]
```diff
@@ -0,0 +1,58 @@
...
+
+        this.client = new analytics.Client({
+            token: token,
+            endpoint: endpoint
+        });
+    }
+
+    send(event: string, customData: any) {
+        if (typeof this.client == 'undefined') {
+            throw new Error(`You must init before sending an event`);
```
Comment (GuiSim) [inline src/SimpleAnalytics.ts]: `You must call init()` ?

File: src/SimpleAnalytics.ts [hunk 33]
```diff
@@ -0,0 +1,58 @@
...
+        }
+
+        customData = objectAssign({}, {
+            hash: window.location.hash
+        }, customData);
+
+        switch (event) {
+            case 'pageview':
```
Comment (GuiSim) [inline src/SimpleAnalytics.ts]: Je me demande si c'est la meilleure façon de faire. Serait pas mieux de faire qqc du genre `analytics.pageView(customData)` ?

File: src/analytics.ts (hunk 15 above)
Comment (GuiSim) [inline src/analytics.ts]: +1 for version in specific param

File: src/analytics.ts [hunk 34]
```diff
@@ -0,0 +1,142 @@
...
+    sendSearchEvent(request: SearchEventRequest): Promise<SearchEventResponse> {
+        return this.sendEvent('search', request).then(defaultResponseTransformer);
+    }
+    sendClickEvent(request: ClickEventRequest): Promise<ClickEventResponse> {
+        return this.sendEvent('click', request).then(defaultResponseTransformer);
+    }
+    sendCustomEvent(request: CustomEventRequest): Promise<CustomEventResponse> {
+        return this.sendEvent('custom', request).then(defaultResponseTransformer);
+    }
+    sendViewEvent(request: ViewEventRequest): Promise<ViewEventResponse> {
```
Comment (GuiSim) [inline src/analytics.ts]: Freaking clean ça. gj

File: src/objectAssign.ts [hunk 35]
```diff
@@ -0,0 +1,40 @@
...
+// https://github.com/sindresorhus/object-assign/blob/master/index.js
+const hasOwnProperty = Object.prototype.hasOwnProperty;
+const getOwnPropertySymbols: any = (<any>Object).getOwnPropertySymbols;
+const propIsEnumerable = Object.prototype.propertyIsEnumerable;
+
+export interface IObjectAssign {
+    (target: any, ...sources: any[]) : any;
+}
+
+const objectAssignPonyfill: IObjectAssign = (target: any, ...sources: any[]): any => {
```
Comment (GuiSim) [inline src/objectAssign.ts]: ponyfill ou polyfill?

Comment (pastjean) [inline src/objectAssign.ts]: ponyfill -> provide same api or standard library
polyfill -> override standard library

File: src/SimpleAnalytics.ts (hunk 33 above)
Comment (pastjean) [inline src/SimpleAnalytics.ts]: Le but de ce wrapper la c'est que le snippet a insérer dans la page des clients reste ultra simple

**Pull request merged**